### PR TITLE
Emit dockerfile.v0 resource in manifest.

### LINKED ISSUE #769: Add docker file manifest resource type
For certain scenarios we want to enable running locally as an executable but when creating the manifest for deployment, expressing the resource as a docker file (for the deployment tool to publish). This will be an annotation that can be added to the resource that the manifest publisher will look at and use instead of the executable if specified.

```C#
// Assumes there's a docker file called Dockerfile in the working directory (btw this name sucks)
var nodeApp = builder.AddExecutable("myapp", "node", "..\\nodeapp", "app.js")
                 .WithServiceBinding(scheme: "http")
                 .AsDockerFileForPublish();
```

```JSON
{
  "catalogservice": {
      "type": "dockerFile.v0",
      "path": "[relative path to]\\Dockerfile",
      "args": [], 
      "context": "[relative path]",
      "env": {
          "ConnectionStrings__postgres": "{postgres.connectionString}"
      },
      "bindings": {
          "http": {
              "scheme": "http",
              "protocol": "tcp",
              "transport": "http",
              "external": false,
              "containerPort": 8080
          }
      }
  }
}
```

cc @mitchdenny @ellismg

### PR DESCRIPTION
**Pull request by mitchdenny**
Fixes #769 

This PR adds an extension method to executable resource builders which flags that when emitted to the manifest they should be emitted as a `dockerfile.v0` resource.

- [x] Basic emission logic for `dockerfile.v0`
- [x] Unit tests for manifest publishing
- [x] Remove changes made to sample (just used for illustration purposes, won't be merged)

Given the following code:

```csharp
var builder = DistributedApplication.CreateBuilder(args);

builder.AddAzureProvisioning();

builder.AddNpmApp("nodeapp", "..\\nodeapp")
    .WithServiceBinding(5555, scheme: "http", env: "PORT")
    .WithDockerfile();
```

The following manifest snippet will be produced:

```json
{
  "resources": {
    "nodeapp": {
      "type": "dockerfile.v0",
      "path": "..\\nodeapp\\Dockerfile",
      "context": "..\\nodeapp",
      "env": {
        "NODE_ENV": "development",
        "PORT": "{bindings.http.port}"
      },
      "bindings": {
        "http": {
          "scheme": "http",
          "protocol": "tcp",
          "transport": "http",
          "containerPort": 5555
        }
      }
    }
}
```

### CODE REVIEW

File: src/Aspire.Hosting/Extensions/ExecutableResourceBuilderExtensions.cs [hunk 1]
```diff
@@ -24,4 +26,27 @@ public static IResourceBuilder<ExecutableResource> AddExecutable(this IDistribut
...
+    {
+        return builder.WithAnnotation(
+            new ManifestPublishingCallbackAnnotation(writer => WriteExecutableAsDockerfileResource(writer, builder.Resource))
+            );
+    }
+
+    private static void WriteExecutableAsDockerfileResource(Utf8JsonWriter writer, ExecutableResource executable)
+    {
+        writer.WriteString("type", "dockerfile.v0");
+        writer.WriteString("path", Path.Combine(executable.WorkingDirectory, "Dockerfile"));
```
Comment (davidfowl) [inline src/Aspire.Hosting/Extensions/ExecutableResourceBuilderExtensions.cs:46]: This needs to be resolved to the relative path using all of the crazy logic in the manifest publisher. Maybe expose a method.

**Comment by mitchdenny** [comment]
@ellismg and @rajeshkamal5050 for thoughts on this from the AZD perspective.

File: samples/eShopLite/AppHost/aspire-manifest.json [hunk 2]
```diff
@@ -1,5 +1,22 @@
...
+      "env": {
+        "NODE_ENV": "development",
+        "PORT": "{bindings.http.port}"
+      },
+      "bindings": {
+        "http": {
+          "scheme": "http",
+          "protocol": "tcp",
+          "transport": "http",
+          "containerPort": 5555
```
Comment (davidfowl) [inline samples/eShopLite/AppHost/aspire-manifest.json:16]: Not sure we should be using this port, but we do need a port. This is the host port today, not the port the container is listening on (which happens to be passed in today).

File: src/Aspire.Hosting/Extensions/ExecutableResourceBuilderExtensions.cs (hunk 1 above)
Comment (davidfowl) [reply src/Aspire.Hosting/Extensions/ExecutableResourceBuilderExtensions.cs:46]: Also, this should blow up if you don't have a dockerfile?

File: samples/eShopLite/nodeapp/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,8 @@
+FROM node
```
Comment (davidfowl) [inline samples/eShopLite/nodeapp/Dockerfile:1]: Make a new sample. Don't force everyone to download node 😄

File: src/Aspire.Hosting/Extensions/ExecutableResourceBuilderExtensions.cs (hunk 1 above)
Comment (mitchdenny) [reply src/Aspire.Hosting/Extensions/ExecutableResourceBuilderExtensions.cs:46]: I think I'm about to make the crazy breaking change to the callback :)

File: samples/eShopLite/AppHost/aspire-manifest.json (hunk 2 above)
Comment (mitchdenny) [reply samples/eShopLite/AppHost/aspire-manifest.json:16]: Ah the new overload tricked me!

File: src/Aspire.Hosting/Extensions/ExecutableResourceBuilderExtensions.cs (hunk 1 above)
Comment (davidfowl) [reply src/Aspire.Hosting/Extensions/ExecutableResourceBuilderExtensions.cs:46]: arg. Can we do it in a non-breaking way? Support both, obsolete the old one?

File: samples/eShopLite/nodeapp/Dockerfile (hunk 3 above)
Comment (mitchdenny) [reply samples/eShopLite/nodeapp/Dockerfile:1]: This isn't staying here -- no one will need to download node ;)

**Comment by davidfowl** [comment]
Since we're making bigger API changes I think we should make more to set out selves up for a better future:
1. Expose an extension method that adds the ManifestCallbackAnnotation (if we had that we could avoid the break here)
2. The ManifestPublisherContext should have the manifest output path as context
3. We should consider flowing the resource itself to the ManifestPublisherContext (closures feel sorta bad).

**Comment by prom3theu5** [comment]
Added support for the dockerfile component to Aspir8 based on the above manifest entry. Will run through some tests over next few days - and follow this PR to see if there are any changes.

Atm - i've implemented it using the ContainerRegistry setup during the `init` command, along with resource name and latest tag on each build.
Tag will be made customizable though, and container registry overridable as a cli option - just not today, i'm a busy bee today 😛

**Comment by mitchdenny** [comment]
> i'm a busy bee today 😛

Appreciate everything you are doing truly. You're making an awesome contribution to the project.

**Comment by DamianEdwards** [comment]
I'm not sure it's super clear that the `.WithDockerFile()` call only applies to the manifest. I look at that and think it's going to build and run a container when launched by the AppHost during local dev. Have we considered making it more explicit?

**Comment by mitchdenny** [comment]
> I'm not sure it's super clear that the `.WithDockerFile()` call only applies to the manifest. I look at that and think it's going to build and run a container when launched by the AppHost during local dev. Have we considered making it more explicit?

Yeah I struggle with this too. The `With` verb doesn't quite work. I'm considering we do something like: `AsDockerfileInManifest(...)` but it seems a little too wordy? Thoughts?

**Comment by DamianEdwards** [comment]
> Yeah I struggle with this too. The With verb doesn't quite work. I'm considering we do something like: AsDockerfileInManifest(...) but it seems a little too wordy? Thoughts?

You know what, I think I prefer that, or even `WithDockerfileInManifest()`

**Comment by mitchdenny** [comment]
I have a weakly held opinion that As is better than With in this case, but I don't want to introduce a new suffix unnecessarily.

**Comment by DamianEdwards** [comment]
> I have a weakly held opinion that As is better than With in this case, but I don't want to introduce a new prefix unnecessarily.

Yep I totally see your point and I could see us adopting the `As` prefix for "transform" operations like this. Do what you feel is best for now :smile:

File: src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+namespace Aspire.Hosting.Publishing;
+
+public sealed class ManifestPublishingContext(string manifestPath, Utf8JsonWriter writer)
+{
+    public string ManifestPath { get; } = manifestPath;
+
+    public Utf8JsonWriter Writer { get; } = writer;
+
+    public string GetManifestRelativePath(string path)
+    {
```
[the remark this answers is not in the document]
Comment (mitchdenny) [reply src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs:15]: Are you sure? The manifest output path should be relative to the current working directory. We then resolve that to a fully qualified path and then compute the relative path between the fully qualified manifest path and the fully qualified project path.

[the remark this answers is not in the document]
Comment (davidfowl) [reply src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs:15]: Not anymore, it's fixed now 😄

**Review by davidfowl — APPROVED** [review]
